File: modulesApp/src/main/java/org/apache/geode/main/MainApp.java
package org.apache.geode.main;

import java.util.ServiceLoader;

import org.apache.geode.service.SampleService;

public class MainApp {
  public static void main(String[] args) {

    var mainApp = new MainApp();

    mainApp.loadImplementationFromServiceLoader(SampleService.class);
    mainApp.testClassLeakage();
  }

  private void loadImplementationFromServiceLoader(Class<SampleService> clazz) {
    ServiceLoader.load(clazz)
        .forEach(sampleService -> System.out.println(sampleService.getValue()));
  }

  private void testClassLeakage() {
    checkClassAndLogError("org.apache.geode.service.mod1.impl.SampleServiceImpl");
    checkClassAndLogError("org.apache.geode.service.mod2.impl.SampleServiceImpl");
    checkClassAndLogError("org.apache.geode.service.mod3.impl.SampleServiceImpl");
    checkClassAndLogError("org.apache.geode.service.mod4.impl.SampleServiceImpl");
    checkClassAndLogError("org.apache.geode.service.SampleService");
    checkClassAndLogError("org.apache.geode.subService.SampleSubService");
    checkClassAndLogError("org.apache.geode.subService.impl.SampleSubServiceImpl");
    checkClassAndLogError("org.springframework.util.StringUtils");
    checkClassAndLogError("org.apache.commons.lang3.StringUtils");
    checkClassAndLogError("com.google.common.base.Strings");
    System.out.println("<< ------------------------ >>");
  }

  private void checkClassAndLogError(String classString) {
    try {
      this.getClass().getClassLoader().loadClass(classString);
      System.out.println("Found and loaded class: " + classString);
    } catch (Exception e) {
      System.err.println("Cannot find: " + classString);
    }
  }
}

File: modulesApp/src/main/java/org/apache/geode/main/MainApp.kt
package org.apache.geode.main

import org.apache.geode.module.service.ModuleService
import org.apache.geode.module.service.impl.JBossModuleServiceImpl
import org.apache.geode.service.SampleService
import org.jboss.modules.Module
import org.jboss.modules.maven.ArtifactCoordinates
import java.lang.reflect.Method

class MainApp(private val moduleService: ModuleService = JBossModuleServiceImpl()) {
    private fun loadImplementationFromServiceLoader(clazz: Class<SampleService>) {
        println("****SampleService Implementations****")
        try {
            val sampleServices = moduleService.loadService(clazz)
            sampleServices.forEach { println(it.value) }
        } catch (e: Exception) {
        }
        println("*************************************")
    }

    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            val mainApp = MainApp()

            mainApp.registerModuleForName("sub-module1")
            mainApp.registerModuleForName("sub-module2")
            mainApp.registerModuleForName("sub-module3")
            mainApp.registerModuleForName("sub-module4")
            mainApp.registerModuleForName("sub-module5")

            val subModule1 = mainApp.loadModule("sub-module1")
            val subModule2 = mainApp.loadModule("sub-module2")
            val subModule3 = mainApp.loadModule("sub-module3")
            val subModule4 = mainApp.loadModule("sub-module4")
            val subModule5 = mainApp.loadModule("sub-module5")

            testClassLeakage(subModule1)
            testClassLeakage(subModule2)
            testClassLeakage(subModule3)
            testClassLeakage(subModule4)
            testClassLeakage(subModule5)

            mainApp.loadImplementationFromServiceLoader(SampleService::class.java)

            val o: Any = mainApp.loadClass("org.apache.geode.subService.impl.DomainObject")!!.newInstance()
            val method: Method = o.javaClass.getMethod("getValue")
            println(method.invoke(o))
        }

        private fun testClassLeakage(module: Module) {
            checkClassAndLogError(module, "org.apache.geode.service.impl.SampleServiceImpl")
            checkClassAndLogError(module, "org.apache.geode.service.SampleService")
            checkClassAndLogError(module, "org.apache.geode.subService.SampleSubService")
            checkClassAndLogError(module, "org.apache.geode.subService.impl.SampleSubServiceImpl")
            checkClassAndLogError(module, "org.springframework.util.StringUtils")
            checkClassAndLogError(module, "org.apache.commons.lang3.StringUtils")
            checkClassAndLogError(module, "com.google.common.base.Strings")
            checkClassAndLogError(module, "org.apache.geode.subService.impl.DomainObject")

            println("<< ------------------------ >>")
        }

        private fun checkClassAndLogError(module: Module, classString: String) {
            try {
                module.classLoader.loadClass(classString)
            } catch (e: Exception) {
                println("${module.name} cannot find $classString")
            }
        }
    }

    private fun loadModule(moduleName: String): Module = moduleService.loadModule(moduleName)

    private fun loadClass(className: String): Class<*>? = moduleService.loadClass(className)

    private fun registerModuleFromJar(coordinates: ArtifactCoordinates, name: String, vararg dependentModules: String) {
        moduleService.registerModuleFromJar(coordinates, name, *dependentModules)
    }

    private fun registerModuleForName(name: String) {
        moduleService.registerModuleForName(name)
    }
}